src/main.cpp

```diff
@@ -2,20 +2,93 @@
 {
     ROS_INFO_STREAM("JACO HARDWARE starting");
     ros::init(argc, argv, "jaco_hardware");
-    ros::NodeHandle nh;
+    ros::NodeHandle nh("~");
 
     JacoRobot robot(nh);
     controller_manager::ControllerManager cm(&robot);
 
     ros::AsyncSpinner spinner(1);
     spinner.start();
 
+    // Run Grav Comp Calibration, writes to "ParametersOptimal_Z.txt"
+    bool runGravCalib = false;
+    nh.getParam("run_grav_calib", runGravCalib);
+    if(runGravCalib) {
+        nh.setParam("run_grav_calib", false);
+        ROS_INFO_STREAM("Running Gravcomp Calibration...");
+        auto params = robot.calcGravcompParams();
+        if (!params.empty()) {
+            std::ostringstream oss;
+            std::copy(params.begin(), params.end()-1,
+                std::ostream_iterator<float>(oss, ","));
+            oss << params.back();
+            ROS_INFO_STREAM("Gravcomp Params: " << oss.str());
+
+            std::string gravCompFile = "ParametersOptimal_Z.txt";
+            if(nh.getParam("grav_comp_file", gravCompFile)) {
+                std::string path = ros::package::getPath("jaco_hardware") 
+                    + "/" + gravCompFile;
+                if(std::rename("ParametersOptimal_Z.txt", path.c_str()) < 0) {
+                    std::perror("Error moving grav comp file to destionation.");
+                    gravCompFile = "ParametersOptimal_Z.txt";
+                }
+            }
+            ROS_INFO_STREAM("Params written to " << gravCompFile);
+            return 0;
+        }
+        return -1;
+    }
+
+    // Zero Torque Sensors
+    bool zeroTorque = false;
+    nh.getParam("zero_torque", zeroTorque);
+    if(zeroTorque) {
+        nh.setParam("zero_torque", false);
+        ROS_INFO_STREAM("Moving Robot To Candlestick [ENTER when ready]...");
+        std::cin.get();
+        bool ret = robot.zeroTorqueSensors();
+        if(!ret) {
+            ROS_ERROR("Could not zero torque sensors");
+            return -1;
+        }
+        return 0;
+    }
+
+    bool enableGravComp = false;
+    nh.getParam("grav_comp", enableGravComp);
+    if(enableGravComp) {
+      nh.setParam("grav_comp", false);
+
+      // Activate default grav comp parameters
+      std::string gravCompFile = "calib/GravComParams_Empty.txt";
+      nh.getParam("grav_comp_file", gravCompFile);
+      ROS_INFO_STREAM("Enabling GravComp with Params: " << gravCompFile);
+      if(robot.useGravcompForEStop(true, gravCompFile)) {
+        ROS_INFO_STREAM("Gravcomp Enabled!");
+
+        // Whether to enter grav comp on start-up
+        bool enterGravComp = false;
+        nh.getParam("start_grav_comp", enterGravComp);
+        if(enterGravComp) {
+            ROS_INFO_STREAM("Entering Grav Comp...");
+            nh.setParam("start_grav_comp", false);
+            robot.enterGravComp();
+        }
+      } // end if(Gravcomp successfuly enabled)
+    } // end if(enableGravComp)
+    nh.setParam("grav_comp_file", "calib/GravComParams_Empty.txt");
```

Comment:
Why do you reset the param to the default here? This seems like it should solely be a user-specified param -- I'm not sure why we would overwrite it.

Imagine a user setting a custom `grav_comp_file` and then running this two times -- they would expect the same behavior, but the second time it will be using default gravComp parameters

Reply:
Modified where all non-sticky params start with `run_` and run extra routines (e.g. calibration) rather than the main routine.

Reply:
That makes sense, yeah I think it is logical for a param that is run_  to get reset since we can't assume users want to run a routine more than once